src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt

```diff
@@ -18,16 +21,51 @@ class CombatSimulation(val world: World) {
             .sortedByDescending { (initiativeRoll, _) -> initiativeRoll }
             .map { (_, eachUnit) -> eachUnit }
 
-    fun run() {
+    fun runRound() {
         initiativeOrder.forEach { unit ->
-            unit.tacticalActionStrategy
-                    .decideTurnActions(world, unit, unit.availableActionOptions)
-                    .let { actionsToExecute ->
-                       unit.executeActions(actionsToExecute)
-                    }
-                    .also {
-                        world.removeUnits(world.findDead())
-                    }
+            if((unit.currentAttributes.getValue(Attribute.CURRENT_HEALTH) as NumericValue).value > 0) {
+                // check if unit alive rather than try removing dead units from list as concurrency risks
+                // caller needs to exception handle this in case CURRENT_HEALTH not provided in json
+                unit.tacticalActionStrategy
+                        .decideTurnActions(world, unit, unit.availableActionOptions)
+                        .let { actionsToExecute ->
+                            unit.executeActions(actionsToExecute)
+                        }
+                        .also {
+                            val deadUnits = world.findDead()
+                            deadUnits.forEach { deadUnit->
+                                println("${deadUnit.name} died")
+                            }
+                            world.removeUnits(deadUnits)
+                        }
+            }
+
         }
     }
+
+    fun runSimulation() {
+        var roundNum = 1
+        while(world.enemyForces.size > 0 && world.friendlyForces.size > 0) {
+            printRoundStartStatus(roundNum)
+            try {
+                runRound()
+            } catch(e: Exception) {
+                println("Error: ${e.message}")
+            }
+
+            roundNum++
+        }
+        if(world.enemyForces.size == 0) {
+            println("Party wins")
+        } else {
+            println("Enemies win")
+        }
+    }
+
+    private fun printRoundStartStatus(roundNum: Int) {
+        println("Start of combat round $roundNum")
+        println("Friendly units alive: ${world.friendlyForces.map{it.name}}")
```

Comment:
[Minor] You should be able to straight up print `${world.friendlyForces}`. It looks a bit more machine-y but I personally prefer it.

Reply:
It's very verbose - contrast the below:
Friendly units alive: [io.hsar.wh40k.combatsimulator.model.UnitInstance@7573e12f, io.hsar.wh40k.combatsimulator.model.UnitInstance@51b1a8f6, io.hsar.wh40k.combatsimulator.model.UnitInstance@2a389173, io.hsar.wh40k.combatsimulator.model.UnitInstance@213860b8]
Enemy units alive: [banditRanged1, banditRanged2, banditRanged3, banditRanged4, banditMelee1, banditMelee2]

Reply:
If we are going to print to stdout at least for the time being as our main output method then I think making it shorter and more human-readable is a better option. It also makes it easier to tie up the units taking actions with the units in the JSON file as you can see the names used in the JSON.

Reply:
Open to ideas?

Reply:
That's a bit of a pain, I guess `UnitInstance` isn't a data class. Yeah, no worries then. If you overrode `toString()` to use `name` that would be a bit more general purpose, that's what I would do but the effect is basically the same.